lib/Target/Lattigo/LattigoEmitter.cpp

```diff
@@ -393,6 +398,235 @@ LogicalResult LattigoEmitter::printOperation(arith::ConstantOp op) {
+    os << resultVarName << " := make([]" << elementType << ", ";
+    os << tensorTy.getNumElements() << ")\n";
+
+    // Now loop and apply the cast
+    os << "for i, val := range " << inputVarName << " {\n";
+    os.indent();
+    os << resultVarName << "[i] = " << elementType << "(val)\n";
+    os.unindent();
+    os << "}\n";
+  } else {
+    auto res = convertType(result.getType());
+    if (failed(res)) {
+      return failure();
+    }
+    std::string resultTy = res.value();
+    os << getName(result) << " := " << resultTy << "(" << inputVarName << ")\n";
+  }
+
+  return success();
+}
+
+LogicalResult LattigoEmitter::printOperation(arith::ExtSIOp op) {
+  return typecast(op.getOperand(), op.getResult());
+}
+
+LogicalResult LattigoEmitter::printOperation(arith::ExtUIOp op) {
+  // Unsigned extension should only be used for booleans.
+  assert(
+      getElementTypeOrSelf(op.getOperand().getType()).getIntOrFloatBitWidth() ==
+          1 &&
+      "expected boolean type for extui");
+
+  // golang cannot directly convert booleans to integers, so we need to branch
+  std::string inputVarName = getName(op.getOperand());
+
+  // If it's a slice, upcast by creating a new slice and looping
+  if (auto tensorTy = dyn_cast<RankedTensorType>(op.getResult().getType())) {
+    if (tensorTy.getRank() > 1) {
+      return op.emitOpError()
+             << "Unsupported input type for extui, expected 1D tensor";
+    }
+    auto res = convertType(tensorTy.getElementType());
+    if (failed(res)) {
+      return failure();
+    }
+    std::string elementType = res.value();
+    // Don't use getName because the usage of the variable is guaranteed
+    std::string resultVarName = variableNames->getNameForValue(op.getResult());
+    os << resultVarName << " := make([]" << elementType << ", ";
+    os << tensorTy.getNumElements() << ")\n";
+
+    // Now loop and apply the cast
+    // for i, val := range input {
+    //   if val {
+    //     result[i] = extendedType(1)
+    //   } else {
+    //     result[i] = extendedType(0)
+    //   }
+    // }
+    os << "for i, val := range " << inputVarName << " {\n";
+    os.indent();
+
+    os << "if val {\n";
+    os.indent();
+    os << resultVarName << "[i] = " << elementType << "(1)\n";
+    os.unindent();
+    os << "} else {\n";
+    os.indent();
+    os << resultVarName << "[i] = " << elementType << "(0)\n";
+    os.unindent();
+    os << "}\n";
+
+    os.unindent();
+    os << "}\n";
+  } else {
+    // A plain if to upcast a bool
+    // var result extendedType
+    // if operand {
+    //   result = extendedType(1)
+    // } else {
+    //   result = extendedType(0)
+    // }
+    auto res = convertType(op.getResult().getType());
+    if (failed(res)) {
+      return failure();
+    }
+    std::string resultName = getName(op.getResult());
+    std::string resultTy = res.value();
+
+    os << "var " << resultName << " " << resultTy << "\n";
+    os << "if " << inputVarName << " {\n";
```

Comment:
I remember reviewing @ZenithalHourlyRate 's recent PR on the lattigo emitter and there's a similar boolean upcast to an integer in the encode operation.

I assume eventually these will converge once those encoding/repition go to the layout, but maybe it's worth tagging it with a TODO.
Mostly this is just repeated code and it reminded me of that other impl of boolToInt64 if/else statement

Reply:
Now is a good a time as ever to generalize.